packages/react-router-dev/typegen/index.ts

```diff
@@ -81,4 +84,91 @@ async function writeAll(ctx: Context): Promise<void> {
     fs.mkdirSync(Path.dirname(typesPath), { recursive: true });
     fs.writeFileSync(typesPath, content);
   });
+
+  const registerPath = Path.join(typegenDir, "+register.ts");
+  fs.writeFileSync(registerPath, register(ctx));
+}
+
+function register(ctx: Context) {
+  const register = ts`
+    import "react-router";
+
+    declare module "react-router" {
+      interface Register {
+        params: Params;
+      }
+    }
+  `;
+
+  const { t } = Babel;
+
+  const typeParams = t.tsTypeAliasDeclaration(
+    t.identifier("Params"),
+    null,
+    t.tsTypeLiteral(
+      Object.values(ctx.config.routes)
+        .map((route) => {
+          // filter out pathless (layout) routes
+          if (route.id !== "root" && !route.path) return undefined;
+
+          const lineage = getRouteLineage(ctx.config.routes, route);
+          const fullpath =
+            route.id === "root"
+              ? "/"
+              : lineage
+                  .map((route) => route.path)
+                  .filter((path) => path !== undefined)
+                  .join("/");
+          const params = parseParams(fullpath);
+          return t.tsPropertySignature(
+            t.stringLiteral(fullpath),
+            t.tsTypeAnnotation(
+              t.tsTypeLiteral(
+                Object.entries(params).map(([param, isRequired]) => {
+                  const property = t.tsPropertySignature(
+                    t.stringLiteral(param),
+                    t.tsTypeAnnotation(t.tsStringKeyword())
+                  );
+                  property.optional = !isRequired;
+                  return property;
+                })
+              )
+            )
+          );
+        })
+        .filter((x): x is Babel.Babel.TSPropertySignature => x !== undefined)
+    )
+  );
+
+  return [register, Babel.generate(typeParams).code].join("\n\n");
+}
+
+function parseParams(fullpath: string) {
+  const result: Record<string, boolean> = {};
+
+  let segments = fullpath.split("/");
+  segments.forEach((segment) => {
+    const match = segment.match(/^:([\w-]+)(\?)?/);
+    if (!match) return;
+    const param = match[1];
+    const isRequired = match[2] === undefined;
+
+    result[param] ||= isRequired;
+    return;
+  });
+
+  const hasSplat = segments.at(-1) === "*";
+  if (hasSplat) result["*"] = true;
+  return result;
+}
+
+function getRouteLineage(routes: RouteManifest, route: RouteManifestEntry) {
+  const result: RouteManifestEntry[] = [];
+  while (route) {
+    result.push(route);
+    if (!route.parentId) break;
+    route = routes[route.parentId];
+  }
+  result.reverse();
+  return result;
```

Comment:
note to self: these are copied from `typegen/generate.ts`, might want to refactor that instead of copy/pasting here